platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongListDisk.java

```diff
@@ -40,29 +94,99 @@ public class LongListDisk extends LongList {
+        final int firstChunkWithDataIndex = (int) (minValidIndex.get() / numLongsPerChunk);
+        final int minValidIndexInChunk = (int) (minValidIndex.get() % numLongsPerChunk);
+
+        // copy the first chunk
+        // can't use tempLongBufferThreadLocal here because it's not initialized yet
+        ByteBuffer transferBuffer = ByteBuffer.allocateDirect(memoryChunkSize);
+        transferBuffer.position(minValidIndexInChunk * Long.BYTES).limit(transferBuffer.capacity());
+        MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
+        transferBuffer.flip();
+        // writing the full chunk, all values before minValidIndexInChunk are zeroes
+        MerkleDbFileUtils.completelyWrite(currentFileChannel, transferBuffer, 0);
+        chunkList.set(firstChunkWithDataIndex, 0L);
+
+        // copy everything except for the first chunk and the last chunk
+        MerkleDbFileUtils.completelyTransferFrom(
+                currentFileChannel, fileChannel, memoryChunkSize, (long) (totalNumberOfChunks - 2) * memoryChunkSize);
+
+        // copy the last chunk
+        transferBuffer.clear();
+        MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
+        transferBuffer.flip();
+        MerkleDbFileUtils.completelyWrite(
+                currentFileChannel, transferBuffer, (long) (totalNumberOfChunks - 1) * memoryChunkSize);
+
+        for (int i = firstChunkWithDataIndex + 1; i < totalNumberOfChunks; i++) {
+            chunkList.set(i, (long) i * memoryChunkSize);
+        }
+    }
+
+    private static Path createTempFile(String sourceFileName) throws IOException {
+        return TemporaryFileBuilder.buildTemporaryDirectory(STORE_POSTFIX).resolve(sourceFileName);
+    }
+
+    @Override
+    protected void onEmptyOrAbsentSourceFile(Path path) throws IOException {
+        Path tempFile = createTempFile(path.toFile().getName());
+        currentFileChannel = FileChannel.open(
+                tempFile, StandardOpenOption.CREATE, StandardOpenOption.READ, StandardOpenOption.WRITE);
     }
 
     /**
      * Stores a long at the given index.
      *
      * @param index the index to use
      * @param value the long to store
-     * @throws IndexOutOfBoundsException if the index is negative or beyond the max capacity of the
-     *     list
+     * @throws IndexOutOfBoundsException if the index is negative or beyond the max capacity of the list
      * @throws IllegalArgumentException if the value is zero
      */
     @Override
     public synchronized void put(final long index, final long value) {
         checkValueAndIndex(value, index);
         try {
-            final ByteBuffer buf = TEMP_LONG_BUFFER_THREAD_LOCAL.get();
-            final long offset = FILE_HEADER_SIZE + (index * Long.BYTES);
+            final ByteBuffer buf = tempLongBufferThreadLocal.get();
+            final long offset = createOrGetChunk(index) + calculateOffsetInChunk(index);
```

Comment:
I am confused here... `LongListDisk` extends `LongList<Long>`, so chunks are individual longs. `createOrGetChunk()` returns a chunk, which is a single long. What's the offset in chunk, then?

Reply:
OK, I think I understand it now... Longs are used just as markers, chunk offsets in the file, but chunks may contain multiple longs, not just one. Still feels confusing.

Reply:
Yes. In case of `LongListDisk` every long in the chunk list signifies a disk offset (start position) of the chunk. And then we do lookups within a chunk by the chunk index. It's not that much different from what's going on in off-heap implementation. The principle is the same.